vstart_runner: refactor and reuse code from qa/cephfs

rishabh-d-dave opened the pull request:
Reduces vstart_runner.py's size by 200 lines so far.

- [x] Refactor FuseMount, KernelMount and CephFSMount
- [x] Refactor LocalFilesystem
- [x] Reuse FuseMount's method in LocalFuseMount
- [x] Reuse KernelMount's method in LocalKernelMount
- [x] Move common methods
	from LocalFuseMount and LocalKernelMount to LocalCephFSMount
	from FuseMount and KernelMount to CephFSMount


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

batrick commented:
@sidharthanup please review this.

rishabh-d-dave commented:
@batrick @sidharthanup ping

sidharthanup commented:
@rishabh-d-dave Please do a rebase.

rishabh-d-dave commented:
Rebased again.

rishabh-d-dave commented:
CI failures look unrelated, re-triggering...

rishabh-d-dave commented:
jenkins test docs

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
jenkins test docs

rishabh-d-dave commented:
jenkins test api

lxbsz commented on qa/tasks/cephfs/fuse_mount.py:
```diff
@@ -173,17 +194,31 @@ def list_connections():
+        if mount_wait > 0:
+            log.info("Fuse mount waits {0} seconds before checking /sys/".format(mount_wait))
+            time.sleep(mount_wait)
+        timeout = int(self.client_config.get('mount_timeout', 30))
```
There maybe no need to get it from the config everytime, save it in `self.timeout` or something else. If the `self.timeout` is not None just return it ?

rishabh-d-dave replied:
Moved. Thanks!

lxbsz commented on qa/tasks/cephfs/fuse_mount.py:
```diff
@@ -58,106 +65,120 @@ def _mount(self, mntopts, check_status):
-        cwd = self.test_dir
+        return mount_cmd
+
+    def _get_nsenter_args(self):
```
Okay I saw you have made it a generic method in the following patch.

rishabh-d-dave replied:
Done.

rishabh-d-dave commented on qa/tasks/cephfs/kernel_mount.py:
```diff
@@ -41,33 +43,43 @@ def mount(self, mntopts=[], createfs=True, check_status=True, **kwargs):
+        log.info('mount command passed')
 
-    def _run_mount_cmd(self, mntopts, check_status):
+    def _get_mount_cmd(self, mntopts):
```
That would be better but passing argument would be strange to read -

```
    @property
    def mount_cmd(self):
        return self._mount_cmd

    @mount_cmd.setter
    def mount_cmd(self, mntopts):

    def _run_mount_cmd(self, mntopts, check_status):
        self.mount_cmd = mntopts # the "strange" line
        mountcmd_stdout, mountcmd_stderr = StringIO(), StringIO()
```

`mount_cmd = self._get_mount_cmd(mntopts)` reads better IMHO.

lxbsz replied:
That's okay for me.

rishabh-d-dave commented on qa/tasks/cephfs/kernel_mount.py:
```diff
@@ -84,27 +96,27 @@ def _run_mount_cmd(self, mntopts, check_status):
-        mountcmd_stdout, mountcmd_stderr = StringIO(), StringIO()
+        return mount_cmd
+
+    def _get_nsenter_args(self):
```
Done.

lxbsz reviewed (approved):
LGTM.

batrick commented:
jenkins test make check

rishabh-d-dave commented:
http://pulpito.front.sepia.ceph.com/rishabh-2020-12-03_18:20:21-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi

rishabh-d-dave commented:
@batrick 
Tests with teuthology passed. Patch for fuse_mount.py and mount.py have been tested. What suite should I run to test kernel_mount,py patch?

batrick commented:
I will include it in my next run this weekend. I will review this today.

rishabh-d-dave commented:
Great, thanks!

rishabh-d-dave commented:
@batrick There's no need for a tracker ticket for this, right?

batrick commented:
> @batrick There's no need for a tracker ticket for this, right?

If it won't be backported, no.

batrick commented:
@rishabh-d-dave commit organization and code changes look good to me. Thanks.

rishabh-d-dave commented:
@batrick Yay, thanks!

batrick commented:
waiting on: https://pulpito.ceph.com/pdonnell-2020-12-08_00:31:36-fs-wip-pdonnell-testing-20201207.222120-distro-basic-smithi/

batrick reviewed (changes requested):
Something broke the fuse client setup with valgrind:

https://pulpito.ceph.com/pdonnell-2020-12-08_00:31:36-fs-wip-pdonnell-testing-20201207.222120-distro-basic-smithi/5691666/

and others. I don't see what's immediately wrong. Suggest you look at past successful runs and compare.

rishabh-d-dave commented:
@batrick Cool

rishabh-d-dave commented:
The following two jobs exercise all changes to `qa/tasks/cephfs/fuse_mount.py`, `qa/tasks/cephfs/kernel_mount.py` and `qa/tasks/cephfs/mount.py` -

https://pulpito.ceph.com/rishabh-2021-02-11_13:58:57-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
https://pulpito.ceph.com/rishabh-2021-02-11_13:27:21-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/

And I ran test_cephfs_shell.TestCD.test_cd_with_args and test_acls using vstart_runner.py to exercise modifications in vstart_runner.py.

rishabh-d-dave commented:
> The following two teuthology jobs exercise all changes to `qa/tasks/cephfs/fuse_mount.py`, `qa/tasks/cephfs/kernel_mount.py` and `qa/tasks/cephfs/mount.py` -
> 
> https://pulpito.ceph.com/rishabh-2021-02-11_13:58:57-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
> https://pulpito.ceph.com/rishabh-2021-02-11_13:27:21-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
> 
> And I ran test_cephfs_shell.TestCD.test_cd_with_args and test_acls using vstart_runner.py to exercise modifications in vstart_runner.py. So, I think this PR is ready for merge.

@full

rishabh-d-dave commented:
> The following two teuthology jobs exercise all changes to `qa/tasks/cephfs/fuse_mount.py`, `qa/tasks/cephfs/kernel_mount.py` and `qa/tasks/cephfs/mount.py` -
> 
> https://pulpito.ceph.com/rishabh-2021-02-11_13:58:57-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
> https://pulpito.ceph.com/rishabh-2021-02-11_13:27:21-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
> 
> And I ran test_cephfs_shell.TestCD.test_cd_with_args and test_acls using vstart_runner.py to exercise modifications in vstart_runner.py. So, I think this PR is ready for merge.

@fullerdj

rishabh-d-dave commented:
@batrick I tested this PR, here are the results - https://github.com/ceph/ceph/pull/37711#issuecomment-777999657. Can you please review/merge this PR? Thanks!

batrick reviewed (changes requested):
Several things:

- Move `73c92f16cd3466f6b325e32ad01cd97c373f376e` to a separate PR.
- Drop `vstart_runner: move few methods definitions`. This is code churn that confuses `git blame` needlessly.
- `vstart_runner: remove duplicate methods from LocalCephManager` should be in a separate PR.
- `qa/cephfs: refactor class LocalFilesystem` moves code lines needlessly, more code churn.
- `qa/cephfs: refactor kernel_mount.py` is too vague about what it's doing. Please update the commit explaining what the goal is. Same for `qa/cephfs: refactor fuse_mount.py`.

rishabh-d-dave commented:
@batrick Thanks for the review.

> Several things:
> 
>     * Move `73c92f16cd3466f6b325e32ad01cd97c373f376e` to a separate PR.

https://github.com/ceph/ceph/pull/39681

>     * Drop `vstart_runner: move few methods definitions`. This is code churn that confuses `git blame` needlessly.

Dropped.
 
>     * `vstart_runner: remove duplicate methods from LocalCephManager` should be in a separate PR.
> 

https://github.com/ceph/ceph/pull/39682

>     * `qa/cephfs: refactor class LocalFilesystem` moves code lines needlessly, more code churn.

Dropped.

>     * `qa/cephfs: refactor kernel_mount.py` is too vague about what it's doing. Please update the commit explaining what the goal is. Same for `qa/cephfs: refactor fuse_mount.py`.

Rewrote the commit message to make it more specific.

rishabh-d-dave commented:
@batrick 

Gave tests a run with teuthology for both FUSE and kernel client; both ran successfully -

http://pulpito.front.sepia.ceph.com/rishabh-2021-02-25_12:24:33-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/

batrick reviewed (changes requested):
This looks good now. Just one oddity I noticed: commit `vstart_runner: reuse KernelMount methods in LocalKernelMount` introduced a `self.Filesystem` that was later deleted. It didn't look necessary, is that right? If so, can you remove it from the git history?

rishabh-d-dave commented:
Ran tests once again just to be sure - http://pulpito.front.sepia.ceph.com/rishabh-2021-03-01_10:36:20-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/.

lxbsz commented on qa/tasks/vstart_runner.py:
```diff
@@ -745,164 +681,85 @@ def _run_python(self, pyscript, py_version='python'):
-        # the PID of the launching process, not the long running ceph-fuse process.  Therefore
-        # we need to give an exact path here as the logic for checking /proc/ for which
-        # asok is alive does not work.
+class LocalKernelMount(LocalCephFSMount, KernelMount):
```
LocalKernelMount is redefined twice.

rishabh-d-dave replied:
FIxed, thanks!

rishabh-d-dave commented:
@batrick ping

Here's our last conversation on this PR - https://github.com/ceph/ceph/pull/37711#issuecomment-785963532

rishabh-d-dave commented:
@batrick 

Tested again, ran successfully -
http://pulpito.front.sepia.ceph.com/rishabh-2021-03-08_14:48:25-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
http://pulpito.front.sepia.ceph.com/rishabh-2021-03-08_14:49:08-fs-wip-rishabh-vr-reuse-AFAP-distro-basic-smithi/
